editor/src/clj/editor/atlas.clj

```diff
@@ -1109,3 +1098,156 @@
                           (< node-child-index (dec (.size children)))))
   (run [selection] (move-node! (selection->image selection) 1)))
+
+(defn- snap-pivot
+  [pivot threshold]
+  (let [snap-values [0.0 0.5 1.0]]
+    (mapv #(or (some (fn [^double snap-value]
+                       (when (< (Math/abs (- % snap-value)) threshold)
+                         snap-value)) snap-values) %) pivot)))
+
+(defn- updated-pivot
+  [node-id ^Vector3d delta]
+  (let [scene (g/node-value node-id :scene)
+        rect (-> scene :renderable :user-data :rect)
+        {:keys [geometry width height]} rect
+        rotated (:rotated geometry)
+        [pivot-x pivot-y] (g/node-value node-id :pivot)
+        delta-x (/ (cond-> (.x delta) rotated -) width)
+        delta-y (/ (cond-> (.y delta) (not rotated) -) height)]
+    (if rotated
+      [(- pivot-x delta-y) (+ pivot-y delta-x)]
+      [(+ pivot-x delta-x) (+ pivot-y delta-y)])))
+
+(defn- pivot-handle
+  [reference-renderable pivot-pos scale]
+  (let [{:keys [geometry x y width height]} (-> reference-renderable :user-data :rect)
+        {:keys [rotated]} geometry
+        [pivot-x pivot-y] pivot-pos
+        absolute-pivot-x (* pivot-x (if rotated height width))
+        absolute-pivot-y (* pivot-y (if rotated width height))
+        x (+ x (if rotated (- width absolute-pivot-y) absolute-pivot-x))
+        y (+ y (- height (if rotated absolute-pivot-x absolute-pivot-y))) 
+        position (-> (mapv #(/ % scale) [x y])
+                     (conj 0.0))]
+    (concat
+     (scene-tools/vtx-add position (scene-tools/vtx-scale [7.0 7.0 1.0] (scene-tools/gen-circle 64)))
+     (scene-tools/vtx-add position (scene-tools/gen-point)))))
+
+(g/defnk produce-manip-delta
+  [original-values manip-space start-action action]
+  (when (and start-action action)
+    (let [{:keys [world-rotation world-transform]} (peek original-values)
+          manip-origin (math/translation world-transform)
+          manip-rotation (scene-tools/get-manip-rotation manip-space world-rotation)
+          lead-transform (doto (Matrix4d.) (.set manip-origin))
+          [start-pos pos] (map #(let [manip-dir (math/rotate manip-rotation (Vector3d. 0.0 0.0 1.0))
+                                      _ (.transform lead-transform manip-dir)
+                                      _ (.normalize manip-dir)
+                                      manip-pos (math/translation lead-transform)]
+                                  (math/line-plane-intersection (:world-pos %) (:world-dir %) (Point3d. manip-pos) manip-dir)) [start-action action])]
+      (doto (Vector3d.) (.sub pos start-pos)))))
+
+(g/defnk produce-renderables [_node-id manip-space camera viewport selected-renderables manip-delta snap-enabled snap-threshold]
+  (if (or (empty? selected-renderables)
+          (second selected-renderables)
+          (not= (-> selected-renderables last :node-id g/node-by-id g/node-type) AtlasImage))
```

Comment:
I worry that `g/node-by-id` is the same as `g/node-value`, meaning that using it should be forbidden from the `produce-` functions. On the other hand, it's guaranteed that the node type of a node does not change. If there is a way to avoid it, I'd prefer us to avoid it...